lib/DDG/Spice/TodayInHistory.pm

```diff
@@ -3,33 +3,83 @@ package DDG::Spice::TodayInHistory;
-    'this day in history',
-);
-
-spice to => 'http://www.history.com/this-day-in-history/rss';
-
-spice wrap_string_callback => 1;
+name "TodayInHistory";
+source "http://wikipedia.com";
+description "List of events which occured on this day in history";
+primary_example_queries "today in history", "this day in history";
+secondary_example_queries "historical events on 27 June";
+category "dates";
+topics "everyday","trivia";
+code_url "https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/lib/DDG/Spice/TodayInHistory.pm";
+attribution github => ["https://github.com/killerfish", "Usman Raza"],
+            twitter => ["https://twitter.com/f1shie", "Usman Raza"];
+
+spice to => 'http://en.wikipedia.org/w/api.php?action=query&titles=$1&prop=revisions&rvlimit=1&rvprop=content&format=json&continue=&redirects&callback={{callback}}';
+
+triggers startend => "today in history";
+triggers start => "historical events on", "this day in history";
+
+# define variables
+my %days = ('1'=>'31','2'=>'29','3'=>'31','4'=>'30','5'=>'31','6'=>'30','7'=>'31','8'=>'31','9'=>'30','10'=>'31','11'=>'30','12'=>'31');
+
+# functions:
+# check and process if we got 2 parameters (month and day)
+sub readParams {
+  # read parameters - do not trigger if invalid parameters are found (delimiters: - / . ' ')
+    my ($parameter1, $parameter2) = split (/[-\/. ]/, $_[0]);
+    if ($parameter2) {
+    my $value;
+        # e.g. "historical events on June 12"
+        return $value if (($value = validateParams($parameter1, $parameter2)) ne 'fail');
+        # e.g. "historical events on 16 Sep"
+        return $value if (($value = validateParams($parameter2, $parameter1)) ne 'fail');
+    }
+    # no parameter - use today date
+    else {
+        return useToday();
+    }
+}
+
+# validate parameters and return day_month on success
+sub validateParams {
+  my $givenDay, my $givenMonth, my $month, my $evalfail;
+    # month name as full name (April), in abbreviated form (Apr) or as 2 digit (04)
+    my @monthPatterns = ('%B', '%b', '%m');
+    for my $monthPattern (@monthPatterns) {
+        $evalfail = 1 if ( eval {
+            my $validDate = Time::Piece->strptime("2000/$_[0]/1", "%Y/".$monthPattern."/%d");
+                $givenDay = 0;
+                $month = $validDate->mon;
+                $givenMonth = $validDate->fullmonth;
+        })
+    }
+    return 'fail' if (!$evalfail);
+    return 'fail' if ($_[1] =~ m/\D+/);
+    return 'fail' if ($_[1] > 31 || $_[1] > $days{$month});
+    if ($_[1] =~ m/^0{1}([1-9]){1}$/) {
+        $givenDay = $1;
+    } else {
+        $givenDay = $_[1];
+    }
+    return $givenDay."_".$givenMonth;
+}
+
+# use today's date if no parameter passed
+sub useToday {
+    my $t = DateTime->now;
+    $t->set_time_zone($loc->time_zone);
+    return $t->mday."_".$t->month_name;
+}
 
 handle remainder => sub {
-    return '' if $_ eq '';
+  # check current date in users timezone
+    if (my $date = readParams($_)) {
```

Comment:
Having more than one space between the day and the month fails as well (it shouldn't):

![screen shot 2015-07-03 at 1 36 51 am](https://cloud.githubusercontent.com/assets/81969/8493151/16090e6a-2124-11e5-9fc2-7bb35465a260.png)